File: cs50/problems/2025/x/sentimental/readability/readability.py
from cs50 import get_string

def calc_idx(L: float, S: float):
    return round(0.0588 * L - 0.296 * S - 15.8)

def is_end_sentence(c: str):
    return c == "!" or c == "?" or c == "."

def parse_text(text: str):
    letters = 0
    words = 0
    sentences = 0

    last = '\0'
    for c in text:
        if(c.isalpha()):
            letters += 1
        elif(c.isspace() and not last.isspace() and not is_end_sentence(last)):
            words += 1
        elif(is_end_sentence(c)):
            sentences += 1
            words += 1
        last = c

    L = letters / words * 100
    S = sentences / words * 100

    return [ L, S ]

def readability():
    text = get_string("Text: ")

    [ l,  s ] = parse_text(text)

    avg = calc_idx(l, s)


    if(avg < 1): return print("Before Grade 1")
    if(avg >= 16): return print("Grade 16+")

    return print(f"Grade {avg}")


readability()

File: cs50/problems/2025/x/readability/readability.c
#import <stdio.h>
#import <cs50.h>
#import <string.h>
#import <ctype.h>
#import <math.h>

int calculate_idx(float l_s[]);
void avg_l_s(string text, float l_s[]);
void print_ans(int idx);
bool end_sentence(char c);
bool end_word(char c);

int main(void) {
  string text = get_string("Text: ");

  float l_s[2];

  avg_l_s(text, l_s);

  int idx = calculate_idx(l_s);

  print_ans(idx);

  return 0;
}

int calculate_idx(float l_s[]) {
  return round(0.0588 * l_s[0] - 0.296 * l_s[1] - 15.8);
}

void print_ans(int idx) {
  if(idx >= 16) {
    printf("Grade 16+\n");
    return;
  }
  if (idx > 1) {
    printf("Grade %d\n", idx);
    return;
  }
  printf("Before Grade 1\n");
}

void avg_l_s(string text, float l_s[]) {
  int letter_cnt = 0;
  int sentence_cnt = 0;

  int words_cnt = 0;

  for(int i = 0; i < strlen(text); i++) {
      if(isalpha(text[i])) {
        letter_cnt++;
        continue;
      }
      if(isspace(text[i]) && !isspace(text[i - 1]) && !end_sentence(text[i - 1])){
       words_cnt++;
       continue;
      }
      if(end_sentence(text[i])) {
        sentence_cnt++;
        words_cnt++;
      }
  }

  float l = ((float) letter_cnt / (float) words_cnt) * 100.0;
  float s = ((float) sentence_cnt / (float) words_cnt) * 100.0;
  l_s[0] = l;
  l_s[1] = s;

}

bool end_sentence(char c) {
  return c == '!' || c == '?' || c == '.';
}
